### PR TITLE
r/codeartifact_repository - new resource

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #13714

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_codeartifact_repository - add new resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCodeArtifactRepository_'
--- PASS: TestAccAWSCodeArtifactRepository_basic (206.26s)
--- PASS: TestAccAWSCodeArtifactRepository_description (247.95s)
--- PASS: TestAccAWSCodeArtifactRepository_upstreams (561.75s)
--- PASS: TestAccAWSCodeArtifactRepository_disappears (174.06s)
```

### CODE REVIEW

**Comment by larstobi** [comment]
@DrFaust92 Thanks for this! Is there any progress on the PR? 👍  :-)

File: website/docs/r/codeartifact_repository.html.markdown [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  repository = "upstream"
+  domain     = aws_codeartifact_domain.test.domain
+}
+
+resource "aws_codeartifact_repository" "test" {
+  repository = "example"
+  domain     = aws_codeartifact_domain.example.domain
+
+  upstream {
+    repository_name = aws_codeartifact_repository.upstream1.repository
```
Comment (ewbankkit) [inline website/docs/r/codeartifact_repository.html.markdown:44]:
```suggestion
    repository_name = aws_codeartifact_repository.upstream.repository
```

File: website/docs/r/codeartifact_repository.html.markdown [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `domain` - (Required) The domain that contains the created repository.
+* `repositoy` - (Required) The name of the repository to create.
```
Comment (ewbankkit) [inline website/docs/r/codeartifact_repository.html.markdown:54]:
```suggestion
* `repository` - (Required) The name of the repository to create.
```

File: website/docs/r/codeartifact_repository.html.markdown [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+### Upstream
+
+* `repository_name` - (Required) The name of an upstream repository.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The Name of Repository.
```
Comment (ewbankkit) [inline website/docs/r/codeartifact_repository.html.markdown:67]:
```suggestion
* `id` - The Name of the Repository.
```

File: website/docs/r/codeartifact_repository.html.markdown [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+* `arn` - The ARN of Repository.
```
Comment (ewbankkit) [inline website/docs/r/codeartifact_repository.html.markdown:68]:
```suggestion
* `arn` - The ARN of the repository.
```

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSCodeArtifactRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCodeArtifactRepository_ -timeout 120m
=== RUN   TestAccAWSCodeArtifactRepository_basic
=== PAUSE TestAccAWSCodeArtifactRepository_basic
=== RUN   TestAccAWSCodeArtifactRepository_owner
=== PAUSE TestAccAWSCodeArtifactRepository_owner
=== RUN   TestAccAWSCodeArtifactRepository_description
=== PAUSE TestAccAWSCodeArtifactRepository_description
=== RUN   TestAccAWSCodeArtifactRepository_upstreams
=== PAUSE TestAccAWSCodeArtifactRepository_upstreams
=== RUN   TestAccAWSCodeArtifactRepository_disappears
=== PAUSE TestAccAWSCodeArtifactRepository_disappears
=== CONT  TestAccAWSCodeArtifactRepository_owner
=== CONT  TestAccAWSCodeArtifactRepository_upstreams
=== CONT  TestAccAWSCodeArtifactRepository_disappears
=== CONT  TestAccAWSCodeArtifactRepository_basic
=== CONT  TestAccAWSCodeArtifactRepository_description
=== CONT  TestAccAWSCodeArtifactRepository_disappears
    resource_aws_codeartifact_repository_test.go:214: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSCodeArtifactRepository_disappears (31.82s)
--- PASS: TestAccAWSCodeArtifactRepository_owner (35.02s)
--- PASS: TestAccAWSCodeArtifactRepository_basic (35.28s)
--- PASS: TestAccAWSCodeArtifactRepository_description (51.38s)
--- PASS: TestAccAWSCodeArtifactRepository_upstreams (71.76s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	71.805s
```

**Comment by breathingdust** [comment]
LGTM! 🚀  Thanks @DrFaust92 

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSCodeArtifactRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCodeArtifactRepository_ -timeout 120m
=== RUN   TestAccAWSCodeArtifactRepository_basic
=== PAUSE TestAccAWSCodeArtifactRepository_basic
=== RUN   TestAccAWSCodeArtifactRepository_owner
=== PAUSE TestAccAWSCodeArtifactRepository_owner
=== RUN   TestAccAWSCodeArtifactRepository_description
=== PAUSE TestAccAWSCodeArtifactRepository_description
=== RUN   TestAccAWSCodeArtifactRepository_upstreams
=== PAUSE TestAccAWSCodeArtifactRepository_upstreams
=== RUN   TestAccAWSCodeArtifactRepository_disappears
=== PAUSE TestAccAWSCodeArtifactRepository_disappears
=== CONT  TestAccAWSCodeArtifactRepository_basic
=== CONT  TestAccAWSCodeArtifactRepository_upstreams
=== CONT  TestAccAWSCodeArtifactRepository_description
=== CONT  TestAccAWSCodeArtifactRepository_owner
=== CONT  TestAccAWSCodeArtifactRepository_disappears
    resource_aws_codeartifact_repository_test.go:214: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSCodeArtifactRepository_disappears (16.94s)
--- PASS: TestAccAWSCodeArtifactRepository_basic (18.52s)
--- PASS: TestAccAWSCodeArtifactRepository_owner (19.41s)
--- PASS: TestAccAWSCodeArtifactRepository_description (31.34s)
--- PASS: TestAccAWSCodeArtifactRepository_upstreams (46.57s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	48.605s
```

**Pull request merged**